### PR TITLE
*: refactor ExecuteInternal to return single resultset (#22546)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #22546 to release-4.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/22640
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/22640:release-4.0-7ca1629d1a51
```

---

### What problem does this PR solve?

Problem Summary:

Both `Execute` and `ExecuteInternal` no longer permit multi-statement, but the return type was maintained the same for backward compatibility.

This changes `ExecuteInternal` to return a single result-set.

**Update:** I've removed `Execute` from this PR, we can handle it in followup PRs.

### What is changed and how it works?

What's Changed:

This changes the return type of `ExecuteInternal` to be a non-array value. ~All uses of `Execute` are changed to use `ExecuteInternal`~ (for future PRs - it does not need to be handled all at once).

### Check List 

Tests 

- Covered by existing unit test

Side effects

- More difficult cherry-picks of code

### Release note 

- No release note

### CODE REVIEW

**Comment by tiancaiamao** [comment]
Wow, so many conflicts...

**Comment by morgo** [comment]
This depends on https://github.com/pingcap/tidb/pull/22548 - I will wait for it to merge first.

**Comment by xhebox** [comment]
But that is blocked on merge. We may need to apply for cherry-picking.

File: bindinfo/handle.go [hunk 1]
```diff
@@ -661,16 +661,16 @@ func (h *BindHandle) CaptureBaselines() {
 func getHintsForSQL(sctx sessionctx.Context, sql string) (string, error) {
 	oriVals := sctx.GetSessionVars().UsePlanBaselines
 	sctx.GetSessionVars().UsePlanBaselines = false
-	recordSets, err := sctx.(sqlexec.SQLExecutor).ExecuteInternal(context.TODO(), fmt.Sprintf("explain format='hint' %s", sql))
+	rs, err := sctx.(sqlexec.SQLExecutor).ExecuteInternal(context.TODO(), fmt.Sprintf("explain format='hint' %s", sql))
 	sctx.GetSessionVars().UsePlanBaselines = oriVals
-	if len(recordSets) > 0 {
-		defer terror.Log(recordSets[0].Close())
+	if rs != nil {
```
Comment (xhebox) [inline bindinfo/handle.go:666]: Same to the 5.0-rc, we can remove the check, and move the `defer` after `err != nil` check.

File: bindinfo/handle.go [hunk 2]
```diff
@@ -873,23 +873,22 @@ func runSQL(ctx context.Context, sctx sessionctx.Context, sql string, resultChan
 			resultChan <- fmt.Errorf("run sql panicked: %v", string(buf))
 		}
 	}()
-	recordSets, err := sctx.(sqlexec.SQLExecutor).ExecuteInternal(ctx, sql)
+	rs, err := sctx.(sqlexec.SQLExecutor).ExecuteInternal(ctx, sql)
 	if err != nil {
-		if len(recordSets) > 0 {
-			terror.Call(recordSets[0].Close)
+		if rs != nil {
```
Comment (xhebox) [inline bindinfo/handle.go:878]: ditto

File: session/bootstrap.go [hunk 3]
```diff
@@ -1160,24 +1160,24 @@ func upgradeToVer51(s Session, ver int64) {
...
 		`SELECT bind_sql, default_db, status, create_time, charset, collation, source
 			FROM mysql.bind_info
 			WHERE source != 'builtin'
 			ORDER BY update_time DESC`)
 	if err != nil {
 		logutil.BgLogger().Fatal("upgradeToVer61 error", zap.Error(err))
 	}
-	if len(recordSets) > 0 {
-		defer terror.Call(recordSets[0].Close)
+	if rs != nil {
```
Comment (xhebox) [inline session/bootstrap.go:1172]: ditto

File: session/session.go [hunk 4]
```diff
@@ -870,37 +870,19 @@ func (s *session) ExecRestrictedSQLWithSnapshot(sql string) ([]chunk.Row, []*ast
...
-		for _, rs := range recordSets {
-			closeErr := rs.Close()
-			if closeErr != nil && err == nil {
-				err = closeErr
-			}
-		}
-	}()
-	if err != nil {
+	rs, err := se.ExecuteInternal(ctx, sql)
+	if rs != nil {
```
Comment (xhebox) [inline session/session.go:874]: ditto

**Review by xhebox — COMMENTED** [review]
Rest LGTM

**Review by AilinKid — COMMENTED** [review]
/lgtm

**Review by AilinKid — APPROVED** [review]

**Pull request merged**